python/tank/bootstrap/manager.py

```diff
@@ -221,11 +244,50 @@ def set_progress_callback(self, callback):
         """
         self._progress_cb = callback
 
-    def bootstrap_engine(self, engine_name, entity=None):
+    def bootstrap_engine(self,
+                         engine_name,
+                         entity=None,
+                         progress_callback=None,
+                         completed_callback=None,
+                         failed_callback=None):
         """
         Create an sgtk instance for the given project or site,
         then launch into the given engine.
 
+        When property ``async_bootstrap`` is false, the whole engine bootstrap logic
+        will be executed synchronously in the main application thread. When true,
+        an :class:`sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+
+        When property ``async_bootstrap`` is true, 3 callback functions should be provided.
+
+        A callback function that reports back on the bootstrap progress
+        with the following signature::
+
+            progress_callback(message, current_index, maximum_index)
+
+        where:
+        - ``message`` is the progress message to report.
+        - ``current_index`` is an optional current item number being looped over.
+        - ``maximum_index`` is an optional maximum item number being looped over.
+
+        A callback function that handles cleanup after successful completion of the bootstrap
+        with the following signature::
+
+            completed_callback(engine)
+
+        where:
+        - ``engine``is the launched :class:`sgtk.platform.Engine` instance.
+
+        A callback function that handles cleanup after failed completion of the bootstrap
+        with the following signature::
+
+            failed_callback(step, exception)
+
+        where:
+        - ``step``is the bootstrap step ("sgtk" or "engine") that raised the exception.
```

Comment:
Could the failed bootstrap roll back to the previous state automatically? Or is that crazy complicated? If it did, we wouldn't need the step option.

If we do keep the step option, I think we should use a constant rather than a string here. I find step is confusing (typically refers to pipeline step in shotgun!) - how about we call it `phase` or `boostrap_phase` instead? Or something like that?

And I imagine most people will be interested in the exception object, so probably better to swap the paramater order too.